x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/common/Truncator.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.inference.common;
+
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.core.Nullable;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Provides truncation logic for inference requests
+ */
+public class Truncator {
+
+    /**
+     * Defines the percentage to reduce the input text for an inference request.
+     */
+    static final Setting<Double> REDUCTION_PERCENTAGE_SETTING = Setting.doubleSetting(
+        "xpack.inference.truncator.reducation_percentage",
+        0.5,
+        0.0,
+        1.0,
+        Setting.Property.NodeScope,
+        Setting.Property.Dynamic
+    );
+
+    public static List<Setting<?>> getSettings() {
+        return List.of(REDUCTION_PERCENTAGE_SETTING);
+    }
+
+    /**
+     * OpenAI estimates that there are 4 characters per token
+     * <a href="https://help.openai.com/en/articles/4936856-what-are-tokens-and-how-to-count-them">here</a>.
+     * We'll take a conservative approach and assume there's a token every 3 characters.
+     */
+    private static final double CHARS_PER_TOKEN = 3;
+
+    public static double countTokens(String text) {
+        return Math.ceil(text.length() / CHARS_PER_TOKEN);
+    }
+
+    private volatile double reductionPercentage;
+
+    public Truncator(Settings settings, ClusterService clusterService) {
+        this.reductionPercentage = REDUCTION_PERCENTAGE_SETTING.get(settings);
+
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(REDUCTION_PERCENTAGE_SETTING, this::setReductionPercentage);
+    }
+
+    private void setReductionPercentage(double percentage) {
+        reductionPercentage = percentage;
+    }
+
+    /**
+     * Truncate each entry in the list to the specified number of tokens.
+     * @param input list of strings
+     * @param tokenLimit the number of tokens to limit the text to
```

Comment:
```suggestion
     * @param charLimit the number of characters to limit the text to
```

The truncator does not tokenize, it is characters rather than tokens

Reply:
We talked about this offline. I believe they can stay because we are passing in the number of tokens and not characters.